python/convert.c

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright (c) 2009-2021, Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of Google LLC nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL Google LLC BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "python/convert.h"
+
+#include "python/protobuf.h"
+#include "upb/reflection.h"
+#include "upb/util/compare.h"
+
+PyObject* PyUpb_UpbToPy(upb_msgval val, const upb_fielddef *f, PyObject *arena) {
+  switch (upb_fielddef_type(f)) {
+    case UPB_TYPE_ENUM:
+    case UPB_TYPE_INT32:
+      return PyLong_FromLong(val.int32_val);
+    case UPB_TYPE_INT64:
+      return PyLong_FromLongLong(val.int64_val);
+    case UPB_TYPE_UINT32:
+      return PyLong_FromSize_t(val.uint32_val);
+    case UPB_TYPE_UINT64:
+      return PyLong_FromUnsignedLongLong(val.uint64_val);
+    case UPB_TYPE_FLOAT:
+      return PyFloat_FromDouble(val.float_val);
+    case UPB_TYPE_DOUBLE:
+      return PyFloat_FromDouble(val.double_val);
+    case UPB_TYPE_BOOL:
+      return PyBool_FromLong(val.bool_val);
+    case UPB_TYPE_BYTES:
+      return PyBytes_FromStringAndSize(val.str_val.data, val.str_val.size);
+    case UPB_TYPE_STRING:
+      return PyUnicode_DecodeUTF8(val.str_val.data, val.str_val.size, NULL);
+    case UPB_TYPE_MESSAGE:
+      PyErr_Format(PyExc_NotImplementedError,
+                   "Conversion of message types not yet implemented");
```

Comment:
can you `return NULL` after this Err as well?

Reply:
Done.